### PR TITLE
Solving #186 Positive reinforcement

### PR DESCRIPTION
**Pull request by shaman-rajan**
I'm still working on editing the data schema to all questions, but created the PR to take comments in the meantime. 2 autosave karma tests fail, and I'm working on fixing that too.

### CODE REVIEW

**Comment by shaman-rajan** [comment]
All tests pass now and all questions have tags included. @eyurko PTAL.

File: assets/tests/TaskSchemaValidationService.js [hunk 1]
```diff
@@ -157,6 +157,12 @@ tie.factory('TaskSchemaValidationService', [
           });
         });
       },
+      verifyCorrectnessTestsHaveTags: function(task) {
+        var correctnessTests = task.getCorrectnessTests();
+        return correctnessTests.every(function(test) {
+          return test.getTag() !== undefined && test.getTag().length > 0;
```
Comment (seanlip) [inline assets/tests/TaskSchemaValidationService.js]: angular.isString(test.getTag()) is probably a stronger check.

File: client/components/LearnerViewDirective.js [hunk 2]
```diff
@@ -481,6 +505,19 @@ tie.directive('learnerView', [function() {
         var autosaveCancelPromise;
         var cachedCode;
         var congratulatoryFeedback = FeedbackObjectFactory.create();
+        var reinforcement = ReinforcementObjectFactory.create();
+        QuestionDataService.initCurrentQuestionSet(questionSetId);
+        var questionSet = QuestionDataService.getCurrentQuestionSet(
+          questionSetId);
+        $scope.currentQuestionIndex = 0;
```
Comment (seanlip) [inline client/components/LearnerViewDirective.js]: Don't copy/paste please -- refactor common stuff into a function instead. That said, I just made a similar comment on #250 which has just been merged, so merging from master into this branch and resolving the conflicts should address this.

File: client/domain/ReinforcementBulletObjectFactory.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+ * reinforcementBullet domain objects.
+ */
+
+tie.factory('ReinforcementBulletObjectFactory', [
+  function() {
+
+    var ReinforcementBullet = function(passed, content) {
+      this._passed = passed;
+      this._content = content;
+      if (passed) {
```
Comment (seanlip) [inline client/domain/ReinforcementBulletObjectFactory.js]: this._imgName = (passed ? 'pass.png' : 'fail.png');

but then, actually, you don't even need this._imgName at all. Just add a method ReinforcementBullet.prototype.getImageFilename() and do the computation there.

File: client/services/ReinforcementGeneratorService.js [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+/**
+ * @fileoverview Service for generating positive reinforcement when a user
+ * submits code based on the tasks that the code passes/fails on.
+ */
+
+tie.factory('ReinforcementGeneratorService', [
+  function() {
+
+    return {
+      getReinforcement: function(task, codeEvalResult, runtimeFeedback) {
```
Comment (seanlip) [inline client/services/ReinforcementGeneratorService.js]: It's an antipattern to add new keys to an object mid-stream -- let's avoid that. It means that the same-looking object has a different definition depending on which file in the code you're looking at.

Instead, construct the Reinforcement object here, and return just that.

File: client/services/SolutionHandlerService.js [hunk 5]
```diff
@@ -60,7 +60,9 @@ tie.factory('SolutionHandlerService', [
               task, codeEvalResult, codeSubmission.getRawCodeLineIndexes());
             TranscriptService.recordSnapshot(
               SnapshotObjectFactory.create(codeEvalResult, runtimeFeedback));
-            return runtimeFeedback;
+            var finalFeedback = ReinforcementGeneratorService.getReinforcement(
```
Comment (seanlip) [inline client/services/SolutionHandlerService.js]: Don't modify the feedback object. Instead, when passing stuff to the learner view directive, pass the reinforcement object should be passed as a separate object from the usual feedback. Do all the computations in services so that there is no logic in the directive controller.

**Comment by shaman-rajan** [comment]
@seanlip PTAL. Comments addressed except the one we can fix during merge.

File: client/components/LearnerViewDirective.js (hunk 2 above)
Comment (shaman-rajan) [inline client/components/LearnerViewDirective.js]: This isn't needed, will remove it.

**Comment by shaman-rajan** [comment]
Looks like there are still issues here. When submitting with a syntax error, the app hangs. Will update once it's fixed.

**Comment by seanlip** [comment]
OK, thanks for the update @shaman-rajan! I'll take another look when it's ready.

**Comment by shaman-rajan** [comment]
@seanlip PTAL. Fixed the syntax error issue, and addressed all comments.

File: client/components/LearnerViewDirective.js [hunk 6]
```diff
@@ -318,6 +324,21 @@ tie.directive('learnerView', [function() {
         .tie-previous-instructions {
           opacity: 0.5;
         }
+        .tie-reinforcement li {
+          list-style: none;
+          margin: 0;
+          margin-top: 1px;
+          position: relative;
+        }
+        .bullet-img {
```
Comment (seanlip) [inline client/components/LearnerViewDirective.js]: Could you use `.tie-bullet-img` or `.tie-reinforcement .bullet-img` instead? We prefix all our CSS classes with `.tie-` in order to distinguish them from third-party libs.

File: client/components/LearnerViewDirective.js [hunk 7]
```diff
@@ -318,6 +324,21 @@ tie.directive('learnerView', [function() {
...
+          width: 15px;
```
Comment (seanlip) [inline client/components/LearnerViewDirective.js]: Nit: recommend alphabetizing these attrs (i.e. bottom, then height, then position, then width). As the number of attributes increases, this makes it easier to locate a particular one.

File: client/components/LearnerViewDirective.js [hunk 8]
```diff
@@ -553,9 +588,28 @@ tie.directive('learnerView', [function() {
             } else if (syntaxErrorIndex === null) {
               $scope.syntaxErrorString = '';
               $scope.syntaxErrorFound = false;
+
+              // Updating reinforcement bullets only if no syntax errors
+              reinforcement.clear();
+              for (var caseList in reinforcementObject.passedList) {
+                if (reinforcementObject.passedList[caseList]) {
```
Comment (seanlip) [inline client/components/LearnerViewDirective.js]: The reinforcement object should be fully prepared in a service and just passed to the controller -- there should be none of this logic in the directive's controller.

If necessary, you can handle the "don't show if there are syntax errors" bit with an if statement -- but that's about as much logic as should exist at this level of the code.

File: client/domain/ReinforcementBulletObjectFactory.js [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    ReinforcementBullet.prototype.isPassedBullet = function() {
+      return this._passed;
+    };
+
+    ReinforcementBullet.prototype.getContent = function() {
+      return this._content;
+    };
+
+    ReinforcementBullet.prototype.getImgName = function() {
+      return this._passed ? "pass.png" : "fail.png";
```
Comment (seanlip) [inline client/domain/ReinforcementBulletObjectFactory.js]: Use single quotes here for consistency with the other files.

File: client/services/ReinforcementGeneratorService.js [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+        }
+        if (!task.pastFailsList) {
+          task.pastFailsList = {};
+        }
+
+        var splitTestsByTag = function(tests) {
+          var splitTests = {};
+          for (var i = 0; i < tests.length; ++i) {
+            var test = tests[i];
+            test.index = i;
```
Comment (seanlip) [inline client/services/ReinforcementGeneratorService.js]: Please avoid modifying dicts like this. If you change the keys midstream then it's hard to tell what a dict's structure is at any given time.

File: client/services/ReinforcementGeneratorService.js [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+tie.factory('ReinforcementGeneratorService', [
+  function() {
+
+    return {
+      getReinforcement: function(task, codeEvalResult) {
+
+        // Initializing question reinforcement data if not done already
+        if (!task.passedList) {
+          task.passedList = {};
```
Comment (seanlip) [inline client/services/ReinforcementGeneratorService.js]: Please avoid modifying dicts like this. If you change the keys midstream then it's hard to tell what a dict's structure is at any given time.

Also, why is this defined on the task? I thought it would be part of the reinforcement object, conceptually?

File: client/services/ReinforcementGeneratorService.js [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+          task.pastFailsList = {};
+        }
+
+        var splitTestsByTag = function(tests) {
+          var splitTests = {};
+          for (var i = 0; i < tests.length; ++i) {
+            var test = tests[i];
+            test.index = i;
+            var testTag = test.getTag();
+            if (testTag in splitTests) {
```
Comment (seanlip) [inline client/services/ReinforcementGeneratorService.js]: Use splitTests.hasOwnProperty(testTag) for checking object membership.

File: client/services/ReinforcementGeneratorService.js [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+        // Initializing question reinforcement data if not done already
+        if (!task.passedList) {
+          task.passedList = {};
+        }
+        if (!task.pastFailsList) {
+          task.pastFailsList = {};
+        }
+
+        var splitTestsByTag = function(tests) {
+          var splitTests = {};
```
Comment (seanlip) [inline client/services/ReinforcementGeneratorService.js]: testsByTag or testsKeyedByTag might be a better name.

File: client/services/ReinforcementGeneratorService.js [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+            if (testTag in splitTests) {
+              splitTests[testTag].push(test);
+            } else {
+              splitTests[testTag] = [test];
+            }
+          }
+          return splitTests;
+        };
+
+        // Go through correctness tests to update reinforcement data
```
Comment (seanlip) [inline client/services/ReinforcementGeneratorService.js]: Add a period at the end of the comment.

File: client/services/SolutionHandlerService.js [hunk 15]
```diff
@@ -39,7 +39,10 @@ tie.factory('SolutionHandlerService', [
               potentialSyntaxErrorString);
             TranscriptService.recordSnapshot(
               SnapshotObjectFactory.create(rawCodeEvalResult, feedback));
-            return $q.resolve(feedback);
+            return {
+              feedbackObject: feedback,
+              reinforcementObject: {}
```
Comment (seanlip) [inline client/services/SolutionHandlerService.js]: This should be a blank Reinforcement object. Avoid passing dicts around.

File: client/services/ReinforcementGeneratorService.js [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+  function() {
+
+    return {
+      getReinforcement: function(task, codeEvalResult) {
+
+        // Initializing question reinforcement data if not done already
+        if (!task.passedList) {
+          task.passedList = {};
+        }
+        if (!task.pastFailsList) {
```
Comment (seanlip) [inline client/services/ReinforcementGeneratorService.js]: I think I understand why you're doing it like this -- you want to maintain history. However, defining stuff on the task dict midstream is not the right way to do this, since the task is supposed to be independent of the individual student's playthrough.

So, there are two ways we can go here:
(a) The "proper" way to do this is probably to preserve info about "past fails" somewhere. Perhaps you can store these in the transcript, and extract them from there.
(b) If you want to get this PR in quicker, though, that may be too much to do. In that case, let's drop the "past fails" requirement for this PR, and handle things in a stateless manner.

I'm happy either way. Let me know if you want to discuss any of this.

Comment (BenjaminHolland) [inline client/services/ReinforcementGeneratorService.js]: I kind of like the idea of having a more robust transcript system, especially for repeat playthroughs. Eventually we might be able to turn that into a progress or improvement system of sorts.

**Comment by seanlip** [comment]
Also, FYI, there are merge conflicts.

File: client/components/LearnerViewDirective.js (hunk 8 above)
Comment (shaman-rajan) [inline client/components/LearnerViewDirective.js]: I was using a style similar to congratulatoryFeedback defined on line 515. The reinforcement object we get from the service is always a new one, from which entries are taken and the main stored object is updated.

Comment (seanlip) [inline client/components/LearnerViewDirective.js]: Ah, I see. Let's not do that -- congratulatoryFeedback is a special case, and is just meant to show a message at the end of the task (which is just a display thing, and no logic is required). However, for something complex like reinforcement data, it's better to just pass the object wholesale from the service and do the logic within the service.

File: client/services/ReinforcementGeneratorService.js (hunk 16 above)
Comment (shaman-rajan) [inline client/services/ReinforcementGeneratorService.js]: I like the idea of adding it to the transcript. Will work on doing that.

**Comment by shaman-rajan** [comment]
@seanlip PTAL

File: client/domain/ReinforcementObjectFactory.js [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+      }
+
+      return bullets;
+    };
+
+    Reinforcement.prototype.getTask = function() {
+      return this._task;
+    };
+
+    Reinforcement.prototype.addToPassedList = function(key, value) {
```
Comment (seanlip) [inline client/domain/ReinforcementObjectFactory.js]: What exactly do `key` and `value` represent here? (E.g. is the "key" a "tag"?)

Please add comments to explain, or use clearer variable names. Ditto below.

File: client/domain/ReinforcementObjectFactory.js [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+              "Handles " + caseList));
+        } else {
+          bullets.push(
+            ReinforcementBulletObjectFactory.createFailedBullet(
+              "Fails " + caseList));
+        }
+      }
+
+      for (var testCase in this._pastFailsList) {
+        if (this._pastFailsList[testCase]) {
```
Comment (seanlip) [inline client/domain/ReinforcementObjectFactory.js]: In principle, is it possible for testCase here to overlap a caseList in this._passedList? If so, how do you guard against that (within this object)?

File: client/services/ReinforcementGeneratorService.js [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+ * submits code based on the tasks that the code passes/fails on.
+ */
+
+tie.factory('ReinforcementGeneratorService', [
+  'ReinforcementObjectFactory', 'TranscriptService',
+  function(ReinforcementObjectFactory, TranscriptService) {
+    return {
+      getReinforcement: function(task, codeEvalResult) {
+
+        var testsKeyedByTag = function(tests) {
```
Comment (seanlip) [inline client/services/ReinforcementGeneratorService.js]: Move this between lines 22 and 23 -- no need for inner functions.

Also: getTestsKeyedByTag. Names of functions should start with verbs.

File: client/services/ReinforcementGeneratorService.js [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+          var splitTests = {};
+          for (var i = 0; i < tests.length; ++i) {
+            var test = tests[i];
+            test.originalIndex = i;
```
Comment (seanlip) [inline client/services/ReinforcementGeneratorService.js]: Please do not modify objects midstream -- we talked about this before.

File: client/services/ReinforcementGeneratorService.js [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+        if (previousSnapshot !== null) {
+          previousReinforcement = previousSnapshot.getReinforcement();
+        }
+
+        var reinforcement = ReinforcementObjectFactory.create(task);
+
+        // Copy the previous reinforcement object if we are on the same task.
+        if (previousReinforcement !== null &&
+            task === previousReinforcement.getTask()) {
+          for (var tag in previousReinforcement.getPassedList()) {
```
Comment (seanlip) [inline client/services/ReinforcementGeneratorService.js]: Define a local variable for previousReinforcement.getPassedList() and use that instead; don't keep calling the method repeatedly (especially in a for loop). Ditto for the past fails list.

It may make sense to add a reinforcement.clonePassedList() method that uses angular.copy() to bulk-populate the list; if you go that route, make sure to document the expected inputs.

File: client/services/ReinforcementGeneratorService.js [hunk 22]
```diff
@@ -0,0 +1,101 @@
...
+              previousReinforcement.getPassedList()[tag]);
+          }
+          for (var pastCase in previousReinforcement.getPastFailsList()) {
+            reinforcement.addToPastFailsList(pastCase,
+              previousReinforcement.getPastFailsList()[pastCase]);
+          }
+        }
+
+        // Go through correctness tests to update reinforcement data.
+        var correctnessTests = testsKeyedByTag(task.getCorrectnessTests());
```
Comment (seanlip) [inline client/services/ReinforcementGeneratorService.js]: correctnessTestsByTag

File: client/services/ReinforcementGeneratorService.js [hunk 23]
```diff
@@ -0,0 +1,101 @@
...
+        observedOutputs = observedOutputs[observedOutputs.length - 1];
+        var failedCaseSeenOverall = false;
+        for (var testTag in correctnessTests) {
+          var failedCaseSeenInTag = false;
+          for (var testIdx = 0; testIdx < correctnessTests[testTag].length;
+              ++testIdx) {
+            var test = correctnessTests[testTag][testIdx];
+            var observedOutput = observedOutputs[test.originalIndex];
+            if (test.matchesOutput(observedOutput)) {
+              if (test.getInput() in reinforcement.getPastFailsList()) {
```
Comment (seanlip) [inline client/services/ReinforcementGeneratorService.js]: reinforcement.hasPastFailure(test.getInput())

File: client/services/ReinforcementGeneratorService.js [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+                reinforcement.addToPastFailsList(test.getInput(), true);
```
Comment (seanlip) [inline client/services/ReinforcementGeneratorService.js]: But it's already in the list. Should this be "modify" instead?

File: client/services/ReinforcementGeneratorService.js [hunk 25]
```diff
@@ -0,0 +1,101 @@
...
+              }
+            } else if (!failedCaseSeenInTag) {
+              // We want to display only 1 new failed case among all tasks.
+              if (!failedCaseSeenOverall) {
+                reinforcement.addToPastFailsList(test.getInput(), false);
+                failedCaseSeenOverall = true;
+              }
+              failedCaseSeenInTag = true;
```
Comment (seanlip) [inline client/services/ReinforcementGeneratorService.js]: I'm confused. If we change failedCaseSeenInTag to true when we see it is false, then what's the point at all of keeping track of failedCaseSeenInTag?

File: client/services/ReinforcementGeneratorService.js [hunk 26]
```diff
@@ -0,0 +1,101 @@
...
+            }
+          }
+          if (failedCaseSeenInTag) {
+            if (testTag in reinforcement.getPassedList()) {
+              reinforcement.addToPassedList(testTag, false);
```
Comment (seanlip) [inline client/services/ReinforcementGeneratorService.js]: I'm starting to think that "passed list" is a bit of a misnomer here. What are your two lists tracking, exactly? Can the names better reflect their contents?

File: client/services/TranscriptService.js [hunk 27]
```diff
@@ -27,6 +27,9 @@ tie.factory('TranscriptService', [
       },
       recordSnapshot: function(snapshot) {
         transcript.recordSnapshot(snapshot);
+      },
+      getPreviousSnapshot: function() {
```
Comment (seanlip) [inline client/services/TranscriptService.js]: getMostRecentSnapshot() perhaps? "Previous" implies any previous one.

File: client/services/ReinforcementGeneratorService.js (hunk 25 above)
Comment (shaman-rajan) [inline client/services/ReinforcementGeneratorService.js]: This is inside an "else if" clause, so is entered only if the first if condition evaluates to false.

Comment (seanlip) [inline client/services/ReinforcementGeneratorService.js]: Ah. Perhaps, then:

    } else {
      failedCaseSeenInTag = true;
      if (!failedCaseSeenOverall) {
        ...
      }
    }

might be slightly less efficient, but clearer?

File: client/domain/ReinforcementObjectFactory.js [hunk 28]
```diff
@@ -21,15 +21,15 @@ tie.factory('ReinforcementObjectFactory', [
   'ReinforcementBulletObjectFactory', function(ReinforcementBulletObjectFactory) {
     var Reinforcement = function(task) {
       this._task = task;
-      this._passedList = {};
-      this._pastFailsList = {};
+      this._passedTagsList = {};
```
Comment (seanlip) [inline client/domain/ReinforcementObjectFactory.js]: One other thing I forgot to mention earlier (sorry) -- can we call these this._passedTags and this._pastFailedCases? They're not lists; lists are ordered.

File: client/domain/ReinforcementObjectFactory.js [hunk 29]
```diff
@@ -59,20 +59,24 @@ tie.factory('ReinforcementObjectFactory', [
...
     };
 
-    Reinforcement.prototype.getPastFailsList = function() {
-      return this._pastFailsList;
+    Reinforcement.prototype.getPastFailedCasesList = function() {
+      return this._pastFailedCasesList;
+    };
+
+    Reinforcement.prototype.hasPastFailedCase = function(caseInput) {
+       return (caseInput in this._pastFailedCasesList);
```
Comment (seanlip) [inline client/domain/ReinforcementObjectFactory.js]: return this._pastFailedCasesList.hasOwnProperty(caseInput)

Btw, are the caseInput's guaranteed to be strings / hashable?

File: client/domain/ReinforcementObjectFactory.js [hunk 30]
```diff
@@ -59,20 +59,24 @@ tie.factory('ReinforcementObjectFactory', [
       return this._task;
     };
 
-    Reinforcement.prototype.addToPassedList = function(key, value) {
-      this._passedList[key] = value;
+    Reinforcement.prototype.addToPassedTagsList = function(key, value) {
```
Comment (seanlip) [inline client/domain/ReinforcementObjectFactory.js]: Still need docstrings.

**Comment by shaman-rajan** [comment]
@seanlip Addressed comments. 1 thing left to consider is whether all inputs are hashable. If we are using templates to enter test cases, they will initially be strings that may later be parsed to other data structures. I will investigate on whether the objects I am using to hash are strings taken from the template or parsed objects and update.

**Comment by shaman-rajan** [comment]
@seanlip PTAL. Addressed all comments including the hashability of inputs.

File: client/domain/CorrectnessTestObjectFactory.js [hunk 31]
```diff
@@ -24,13 +24,21 @@ tie.factory('CorrectnessTestObjectFactory', [
...
     };
 
     // Instance methods.
     CorrectnessTest.prototype.getInput = function() {
       return this._input;
     };
 
+    CorrectnessTest.prototype.getStringifiedInput = function() {
+      if (!this._stringifiedInput) {
+        this._stringifiedInput = JSON.stringify(this._input);
```
Comment (seanlip) [inline client/domain/CorrectnessTestObjectFactory.js]: Again, do not modify dicts mid-stream. Either always compute on the fly, or store at the outset and do a simple lookup.

File: client/domain/ReinforcementObjectFactory.js [hunk 32]
```diff
@@ -0,0 +1,104 @@
...
+ * @fileoverview Factory for creating new frontend instances of Reinforcement
+ * domain objects.
+ */
+
+tie.factory('ReinforcementObjectFactory', [
+  'ReinforcementBulletObjectFactory', function(
+      ReinforcementBulletObjectFactory) {
+    var Reinforcement = function(task) {
+      this._task = task;
+      this._passedTags = {};
```
Comment (seanlip) [inline client/domain/ReinforcementObjectFactory.js]: Can you document the keys/values (and their types) for these two objects?

File: client/domain/ReinforcementObjectFactory.js [hunk 33]
```diff
@@ -0,0 +1,104 @@
...
+
+      return bullets;
+    };
+
+    Reinforcement.prototype.getTask = function() {
+      return this._task;
+    };
+
+    // Add a tag to the list of tags the code passes/fails
+    Reinforcement.prototype.addToPassedTags = function(tag, passedBool) {
```
Comment (seanlip) [inline client/domain/ReinforcementObjectFactory.js]: addPassedTag

File: client/domain/ReinforcementObjectFactory.js [hunk 34]
```diff
@@ -0,0 +1,104 @@
...
+      this._passedTags[tag] = passedBool;
+    };
+
+    // Update the pass/fail status of a tag
+    Reinforcement.prototype.updatePassedTags = function(tag, passedBool) {
```
Comment (seanlip) [inline client/domain/ReinforcementObjectFactory.js]: updatePassedTag

File: client/domain/ReinforcementObjectFactory.js (hunk 33 above)
Comment (seanlip) [inline client/domain/ReinforcementObjectFactory.js]: Also, here and below, would testPassed be a better name for the boolean? In general you'll notice that the variable/arg names don't include type information. It seems like a good idea to point out that this is a boolean but I would recommend doing it in the comment/docstring.

File: client/domain/ReinforcementObjectFactory.js [hunk 35]
```diff
@@ -0,0 +1,104 @@
...
+      this._passedTags[tag] = passedBool;
+    };
+
+    // Add a test case to the list of failed test cases
+    Reinforcement.prototype.addToPastFailedCases = function(
+        caseInput, passedBool) {
```
Comment (seanlip) [inline client/domain/ReinforcementObjectFactory.js]: The comments here and below should indicate that caseInput is a string, or perhaps you can rename to stringifiedInput.

File: client/domain/ReinforcementObjectFactory.js [hunk 36]
```diff
@@ -0,0 +1,104 @@
...
+    Reinforcement.prototype.getPassedTags = function() {
+      return this._passedTags;
+    };
+
+    Reinforcement.prototype.getPastFailedCases = function() {
+      return this._pastFailedCases;
+    };
+
+    Reinforcement.prototype.hasPastFailedCase = function(caseInput) {
+      return (caseInput in this._pastFailedCases);
```
Comment (seanlip) [inline client/domain/ReinforcementObjectFactory.js]: return this._pastFailedCases.hasOwnProperty(stringifiedInput);

File: client/services/ReinforcementGeneratorService.js [hunk 37]
```diff
@@ -0,0 +1,120 @@
...
+ * submits code based on the tasks that the code passes/fails on.
+ */
+
+tie.factory('ReinforcementGeneratorService', [
+  'ReinforcementObjectFactory', 'TranscriptService',
+  function(ReinforcementObjectFactory, TranscriptService) {
+
+    var getTestsKeyedByTagAndIndexes = function(tests) {
+      var splitTests = {};
+      var inputToOriginalIndex = {};
```
Comment (seanlip) [inline client/services/ReinforcementGeneratorService.js]: I think it's worth adding comments to explain what the keys/values of these two objects are, and why they're needed. The names are reasonable but a bit hard to parse for someone who doesn't have context.

File: client/services/ReinforcementGeneratorService.js [hunk 38]
```diff
@@ -0,0 +1,120 @@
...
+
+        var reinforcement = ReinforcementObjectFactory.create(task);
+
+        // Copy the previous reinforcement object if we are on the same task.
+        if (previousReinforcement !== null &&
+            task === previousReinforcement.getTask()) {
+          var previousReinforcementPassedTags =
+            previousReinforcement.getPassedTags();
+          for (var tag in previousReinforcementPassedTags) {
+            reinforcement.addToPassedTags(tag,
```
Comment (seanlip) [inline client/services/ReinforcementGeneratorService.js]: Here and below: if a line of code spans multiple lines, break after "(".

File: client/services/ReinforcementGeneratorService.js [hunk 39]
```diff
@@ -0,0 +1,120 @@
...
+        inputToOriginalIndex[test.getStringifiedInput()] = i;
+        var testTag = test.getTag();
+        if (splitTests.hasOwnProperty(testTag)) {
+          splitTests[testTag].push(test);
+        } else {
+          splitTests[testTag] = [test];
+        }
+      }
+      return {
+        splitTests: splitTests,
```
Comment (seanlip) [inline client/services/ReinforcementGeneratorService.js]: Avoid returning composite blobs like this where possible. In this case, have two separate methods instead, each of which only returns one thing.

File: client/services/ReinforcementGeneratorService.js [hunk 40]
```diff
@@ -0,0 +1,120 @@
...
+        }
+
+        // Go through correctness tests to update reinforcement data.
+        var correctnessTestsByTagAndIndexes = getTestsKeyedByTagAndIndexes(
+          task.getCorrectnessTests());
+        var correctnessTestsByTag = correctnessTestsByTagAndIndexes.splitTests;
+        var testIndexes = correctnessTestsByTagAndIndexes.testIndexes;
+        var observedOutputs = codeEvalResult.getCorrectnessTestResults();
+        // Get results only for the last task.
+        observedOutputs = observedOutputs[observedOutputs.length - 1];
```
Comment (seanlip) [inline client/services/ReinforcementGeneratorService.js]: Don't redefine a variable. Instead, create a new method along the lines of codeEvalResult.getLastTaskResults()

File: client/domain/CorrectnessTestObjectFactory.js (hunk 31 above)
Comment (shaman-rajan) [inline client/domain/CorrectnessTestObjectFactory.js]: I didn't want to store that unless this function is used. That said, I can do this computation during object creation also.

Comment (seanlip) [inline client/domain/CorrectnessTestObjectFactory.js]: Let's either always do it at object creation time, or always compute it on the fly. I'm fine either way.

**Comment by shaman-rajan** [comment]
@seanlip PTAL

File: client/domain/CorrectnessTestObjectFactory.js [hunk 41]
```diff
@@ -22,15 +22,21 @@ tie.factory('CorrectnessTestObjectFactory', [
...
+      this._tag = correctnessTestDict.tag;
     };
 
     // Instance methods.
     CorrectnessTest.prototype.getInput = function() {
       return this._input;
     };
 
+    CorrectnessTest.prototype.getStringifiedInput = function() {
+      return JSON.stringify(this._input);
```
Comment (seanlip) [inline client/domain/CorrectnessTestObjectFactory.js]: If you are storing the stringified input, and if the input doesn't change, why are you redoing the serialization?

In other words, either do this, or keep line 25, but not both.

File: client/domain/ReinforcementObjectFactory.js [hunk 42]
```diff
@@ -0,0 +1,120 @@
...
+    *  task: The corresponding task for the reinforcement comments.
+    *  passedTags: Dictionary containing tags (test sets) that the code
+    *              passed/failed.
+    *              Key: The tag
+    *              Value: Boolean indicating whether the code passed the tag.
+    *  pastFailedCases: The list of test cases that will be displayed in the
+    *                   reinforcement comments and the corresponding pass/fail
+    *                   status.
+    *                   Key: Stringified input of test case.
+    *                   Value: Boolean indicating where the code passed the
```
Comment (seanlip) [inline client/domain/ReinforcementObjectFactory.js]: where --> whether?

File: client/domain/ReinforcementObjectFactory.js [hunk 43]
```diff
@@ -0,0 +1,120 @@
...
+tie.factory('ReinforcementObjectFactory', [
+  'ReinforcementBulletObjectFactory', function(
+      ReinforcementBulletObjectFactory) {
+    /**
+    * The reinforcement object contains the current state of reinforcement
+    * comments about a particular task in a question.
+    * Fields:
+    *  task: The corresponding task for the reinforcement comments.
+    *  passedTags: Dictionary containing tags (test sets) that the code
+    *              passed/failed.
```
Comment (seanlip) [inline client/domain/ReinforcementObjectFactory.js]: Add: A tag is considered "failed" if at least one of the correctness tests corresponding to the tag have failed.

File: client/domain/ReinforcementObjectFactory.js [hunk 44]
```diff
@@ -0,0 +1,120 @@
...
+            ReinforcementBulletObjectFactory.createPassedBullet(
+              "Handles " + tag));
+        } else {
+          bullets.push(
+            ReinforcementBulletObjectFactory.createFailedBullet(
+              "Fails " + tag));
+        }
+      }
+
+      for (var testCase in this._pastFailedCases) {
```
Comment (seanlip) [inline client/domain/ReinforcementObjectFactory.js]: var stringifiedTestCase

Recommend similar changes elsewhere throughout this file (and possibly others), if necessary, so that it's always clear whether we're dealing with a stringified version or the raw version.

**Comment by shaman-rajan** [comment]
@seanlip PTAL

**Comment by seanlip** [comment]
@shaman-rajan -- this looks great to me, and I'm going to merge it. Thank you very much for all your work on this!
